lib/googleauth/id_tokens/key_sources.rb

```diff
@@ -0,0 +1,394 @@
+        #
+        def refresh_keys
+          @monitor.synchronize do
+            return @current_keys if Time.now < @allow_refresh_at
+            @allow_refresh_at = Time.now + @retry_interval
+
+            response = Net::HTTP.get_response uri
+            raise KeySourceError, "Unable to retrieve data from #{uri}" unless response.is_a? Net::HTTPSuccess
+
+            data = begin
+                     JSON.parse response.body
+                   rescue JSON::ParserError
+                     raise KeySourceError, "Unable to parse JSON"
+                   end
+
+            @current_keys = Array(interpret_json(data))
+          end
+        end
+
+        protected
+
+        def interpret_json
+          nil
+        end
+      end
+
+      ##
+      # A key source that downloads X509 certificates.
+      # Used by the legacy OAuth V1 public certs endpoint.
+      #
+      class X509CertHttpKeySource < HttpKeySource
+        ##
+        # Create a key source that downloads X509 certificates.
+        #
+        # @param uri [String,URI] The URI from which to download keys.
+        # @param algorithm [String] The algorithm to use for signature
+        #     verification. Defaults to "`RS256`".
+        # @param retry_interval [Integer,nil] Override the retry interval in
+        #     seconds. This is the minimum time between retries of failed key
+        #     downloads.
+        #
+        def initialize uri, algorithm: "RS256", retry_interval: nil
+          super uri, retry_interval: retry_interval
+          @algorithm = algorithm
+        end
+
+        protected
+
+        def interpret_json data
+          data.map do |id, cert_str|
+            key = OpenSSL::X509::Certificate.new(cert_str).public_key
+            KeyInfo.new id: id, key: key, algorithm: @algorithm
+          end
+        rescue OpenSSL::X509::CertificateError
+          raise KeySourceError, "Unable to parse X509 certificates"
+        end
+      end
+
+      ##
+      # A key source that downloads a JWK set.
+      #
+      class JwkHttpKeySource < HttpKeySource
+        ##
+        # Create a key source that downloads a JWT Set.
+        #
+        # @param uri [String,URI] The URI from which to download keys.
+        # @param retry_interval [Integer,nil] Override the retry interval in
+        #     seconds. This is the minimum time between retries of failed key
+        #     downloads.
+        #
+        def initialize uri, retry_interval: nil
+          super uri, retry_interval: retry_interval
+        end
+
+        protected
+
+        def interpret_json data
+          KeyInfo.from_jwk_set data
+        end
+      end
+
+      ##
+      # A key source that aggregates other key sources. This means it will
+      # aggregate the keys provided by its constituent sources. Additionally,
+      # when asked to refresh, it will refresh all its constituent sources.
+      #
+      class AggregateKeySource
```

Comment:
My original use case for this was a single verifier that knows how to verify both normal federated signin tokens and IAP tokens (because it aggregates keys from both URLs). And indeed, earlier I had a single `IDTokens.verify` convenience method that used this aggregator. But then later I split that method out into separate `IDTokens.verify_oidc` and `IDTokens.verify_iap`, so that code path is no longer using this class.

I think the use case is still valid if the user wants to use this in a more general manner beyond Google's certs. For example, use the Google public certs, but also try this hard-coded custom key (i.e. use the union of a JwkHttpKeySource for the Google certs, and a StaticKeySource for the hard-coded key). If you don't think those are reasonable use cases for the library to support, we can drop this class, as it's not critical for any of the core use cases.

As for what would happen if duplicate key IDs are used: Currently the library makes the key ID available as a field, but doesn't actually use it, nor does it care if there are duplicates. If two sources with duplicate key IDs are downloaded, all keys from both sources are still tried against a token, regardless of their key IDs. However, a user could write additional logic to filter against key IDs if they want.

Reply:
Makes sense! I don't have an issue keeping it if you think it's useful. Also I think splitting out the verify method into the two separate methods is great, as trying an ID token against all the public keys from both sources every time is a bit inefficient.